book/animations.md

```diff
@@ -638,226 +634,321 @@ trigger it once every 2 seconds:
-        try_numeric_animation(node, "width",
-            old_style, node.style, tab, is_px=True)
-```
-
-Implement the `try_numeric_animation` function. It will compare the
-values of a CSS property in the old and new style. If they are different *and*
-there is a CSS transition defined for those properties, an animation will
-begin. Start with a helper method that returns the duration (in units of animation
-frame count) of a transition if it was set, and `None` otherwise. This requires
-parsing the comma-separated `transition` syntax.^[Unfortunately, setting up
-animations tends to have a lot of boilerplate code, so get ready for more code
-than usual. The good news though is that it's all pretty simple to
-understand.]
+[^more-units]: In real browsers, there are a [lot more][units] units
+to contend with. And other constraints, too---like the fact that
+opacity should be a value between 0 and 1.
+
+[units]: https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Values_and_units
+
+Now add a frame count and an `animate` method that increments it:
 
 ``` {.python}
-def get_transition(property_name, style):
-    if not "transition" in style:
-        return None
-    transition_items = style["transition"].split(",")
-    found = False
-    for item in transition_items:
-        if property_name == item.split(" ")[0]:
-            found = True
-            break
-    if not found:
-        return None   
-    duration_secs = float(item.split(" ")[1][:-1])
-    return duration_secs / REFRESH_RATE_SEC 
+class NumericAnimation:
+    def __init__(self, old_value, new_value, num_frames):
+        # ...
+        self.frame_count = 1
+
+    def animate(self):
+        self.frame_count += 1
+        if self.frame_count >= self.num_frames: return
+```
+
+If the animation is animating, we'll need to compute the new value of
+the property. Let's return that from the `animate`
+method:[^animation-curve]
+
+[^animation-curve]: Note that this class implements a linear animation
+interpretation (or *easing function*). By default, real browsers use a
+non-linear easing function, which looks better, so the demos from this
+chapter will not look quite the same in your browser.
+
+``` {.python}
+class NumericAnimation:
+    def __init__(self, old_value, new_value, num_frames):
+        # ...
+        total_change = self.new_value - self.old_value
+        self.change_per_frame = total_change / num_frames
+
+    def animate(self):
+        # ...
+        current_value = self.old_value + \
+            self.change_per_frame * self.frame_count
+        if self.is_px:
+            return "{}px".format(current_value)
+        else:
+            return "{}".format(current_value)
 ```
 
-Add some code that detects if a transition remains across a style
-recalc, by comparing two style objects---the ones before and after a style
-update for a DOM node:
+Here I've chosen to compute `change_per_frame` in the constructor. Of
```

Comment:
Is this important enough to put in the main text?

Reply:
Doesn't have to be there. I thought it was interesting to point out that some pre-computation could happen here, but now that I think about it a footnote seems more appropriate.